doc/designs/quic-design/rx-depacketizer.md

```diff
@@ -0,0 +1,249 @@
+     * and |packet_payload_offset| if the offset to the packet payload if
+     * there is any.
+     *
+     * When passing this structure from the TX packetizer to the QUIC Read
+     * Record Layer, |packet_data| is expected to only hold the packet
+     * header, and |frames| holds the rest of the data.  It's up to the QUIC
+     * Read Record Layer to encrypt and send all those chunks of data in a
+     * manner fitting for the network medium with as little copying as
+     * possible.
+     *
+     * When passing this structure to the RX depacketizer from the QUIC Read
+     * Record Layer, |packet_data| is expected to contain the whole packet,
+     * and |frames| to be NULL.
+     */
+    unsigned char *packet_data;
+    size_t packet_data_length;              /* Only used for RX purposes */
+    size_t packet_header_length;
+    size_t packet_payload_offset;           /* Only used for RX purposes */
+
+    STACK_OF(OSSL_QUIC_FRAME) *frames;      /* Only used for TX purposes */
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+
+    /* Packet types */
+    enum {
+        pkt_initial,
+        pkt_handshake,
+        pkt_rtt0,
+        pkt_rtt1,
+        pkt_retry
+    } packet_type;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+Interactions
+------------
+
+The RX depacketizer receives a packet from an QUIC Read Record Layer, and
+the processes frames in two phases:
+
+1.  [Collect information for the ACK Manager](#collect-information-for-the-ack-manager)
+2.  [Pass frame data](#pass-frame-data)
+
+then loops a second time through the frames to sends the data from each
+frame to diverse other components, depending on the frame type.
+
+### Other components
+
+There are a number of other components that the RX depacketizer wants to
+interact with:
+
+-   [ACK manager]
+-   Handshake manager, which is currently unspecified.  In the [overview],
+    it's called the "TLS Handshake Record Layer", but that may need some
+    wrapping to fit the purpose of a Handshake manager.
+-   Session manager, which is currently unspecified for QUIC, but may very
+    well be the existing `SSL_SESSION` functionality, extended to fit QUIC
+    purposes.
+-   Flow control, which is currently unspecified.  In the [overview], it's
+    called the "Flow Controller And Statistics Collector"
+-   Connection manager, which is currently unspecified.  In the [overview],
+    there's a "Connection State Machine" that the "RX Frame Handler" isn't
+    talking directly with, so it's possible that the Connection manager will
+    turn out to be the Handshake manager.
+-   Stream SSL objects, to pass the stream data to.
+
+### Receiving data
+
+To pass a packet to the RX depacketizer, use this functions:
+
+``` C
+__owur int ossl_quic_depacketise(const OSSL_QUIC_PACKET *packet,
```

Comment:
I see this the other way around, i.e. the RX packetiser "pulling" data to read rather than it being pushed into it. I expect there to still be a function to say "depacketise now" - but it would not include the packet data. Instead when that function is called the depacketiser would ask the read record layer "please give me a packet to process".

Reply:
I'm not convinced, but have rewritten this section to include both the push and the pull model.